### PR TITLE
Added support for sequential read-ahead file

### PR DESCRIPTION
**Pull request by elipoz**

### CODE REVIEW

File: util/file_reader_writer.cc [hunk 1]
```diff
@@ -743,13 +735,147 @@ class ReadaheadRandomAccessFile : public RandomAccessFile {
...
+  const std::unique_ptr<RandomAccessFile> file_;
+  const size_t alignment_;
+  const size_t readahead_size_;
+
+  mutable std::mutex lock_;
+  mutable AlignedBuffer buffer_;
+  mutable uint64_t buffer_offset_;
+};
+
+class ReadaheadSequentialFile : public SequentialFile {
```
[the remark this answers is not in the document]
Comment (riversand963) [reply util/file_reader_writer.cc]: Will take a look @siying

[the remark this answers is not in the document]
Comment (elipoz) [reply util/file_reader_writer.cc]: @siying I am not sure I understand the proposal. 

I followed the pattern of NewReadaheadRandomAccessFile, which returns RandomAccessFile, which can later be used with RandomAccessFileReader.

Similarly, we now have NewReadaheadSequentialFile, which returns SequentialFile, which can be used with the existing SequentialFileReader (assuming use_direct_io is false, as otherwise it fallback to PositionedRead). All stats, rate limiting, etc. can be added to the existing SequentialFileReader.

So how is your proposal different from what I already have?

**Comment by siying** [comment]
Btw, we are working hard towards removing `ReadaheadRandomAccessFile`, and we are now close. This has created several problems in the past. I don't want to make the same mistake again with sequence file.

File: util/file_reader_writer.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (siying) [reply util/file_reader_writer.cc]: I would create a different class, rather than a class inheriting SequenceFile. That's it:) Maybe it's as simple as remove the "public SequenceFile" and change the class name in some places when it is used.

[the remark this answers is not in the document]
Comment (siying) [reply util/file_reader_writer.cc]: @elipoz sorry I didn't see this comment when I posted the other comment. I hope we can remove ReadaheadRandomAccessFile. The fact that it is a wrapper and implemented RandomAccessFile caused several problems in the past. One instance I remembered was that we changed the meaning of a function in RandomAccessFile() and forgot that it would break ReadaheadRandomAccessFile as it also implements the function.

[the remark this answers is not in the document]
Comment (elipoz) [reply util/file_reader_writer.cc]: Hmmm... Stacking functionality without changing interface is a common design pattern in C++...

The SequentialFileReader class already exists. Should I add a new class - ReadaheadSequentialFileReader (that doesn't inherit anything)? Or should I pack all the read-ahead logic into existing SequentialFileReader?

Either approach seems a bit awkward since the former creates plurality of XXXFileReader classes (which do not have common interface, if we do not use virtual methods), while the latter creates a class with too much logic in it that is hard to follow, having to deal with read-ahead, direct-io, and other features (e.g., it already has ROCKSDB_LITE codepath).

Please advise, since I do not see which one is better.

[the remark this answers is not in the document]
Comment (elipoz) [reply util/file_reader_writer.cc]: "readahead logic to there will be the best solution" -  replied to this already. There are certain complexities to it. See above.
Anyway, the latest version hides the creation of new ReadaheadSequenceFile.

**Comment by elipoz** [comment]
due to mess with git i am moving the changes to https://github.com/facebook/rocksdb/pull/5580
